### PR TITLE
Layer-shell support

### PR DESCRIPTION
**Pull request by masozzi**
I have changed the protocol from wf-shell to layer-shell.
This has only been test on sway, i still need to test i on wayfire.

problems found so far:
* with layer-shell there is no way (that i know of) to set the coordinate on the screen for the virtual keyboard, so for the moment the keyboard is in the middle of the screen. Any idea on how to handle this? should we just anchor the keyboard to the bottom?
* as of now on sway there is no way to close the keyboard, because the keyboard interactivity is set to false.

Apart for those two anoying things, everything else seems to work.

### CODE REVIEW

**Comment by AdrianVovk** [comment]
This looks cool, but it might be worth using [gtk-layer-shell](https://github.com/wmww/gtk-layer-shell) to make it easier to maintain

**Comment by masozzi** [comment]
I totaly forgot about that... i'll give it a try.
Thanks!

**Comment by ammen99** [comment]
@masozzi The two things you outlined, plus the fact that you can't interactively drag a window with layer-shell, is the reason why I use wf-shell. Do you think we could have both backends? That way, the keyboard will work on sway, and it will continue working with the full feature set in wayfire.

**Comment by masozzi** [comment]
> Do you think we could have both backends?

yes, i think we could have both.
although, i think managing the keyboard options for both wf-shell and layer-shell will become a little messy; we would end up with some options that would work just for wf-shell and other just for layer-shell

**Comment by ammen99** [comment]
@masozzi What options would work for just wf-shell and just for layer-shell?
Anyway, i would be fine with it.

**Comment by masozzi** [comment]
For example: x and y coordinate are just for wf-shell.
For layer-shell, we need to add the option to change the default anchor. Optionally we can add the margin.

**Comment by ammen99** [comment]
wf-shell also supports positioning via anchors and margins.

**Comment by masozzi** [comment]
@ammen99 Sorry if took long.

so, i've added _clara_ which is a hendy helper for parsing command line options;
and the support for anchors.
also, wf-shell is back alogn side with layer-shell.

let me know

**Comment by ammen99** [comment]
@masozzi, I'll do an in-depth review later. At first glance though, do we need clara when we have `getopt_long`? What does it let us achieve we can't otherwise?

**Comment by masozzi** [comment]
@ammen99 
> do we need clara when we have getopt_long?

honestly is just to have a cleaner code, and easier to add new options (if needed).
I can got for getopt_long if you prefer, let me know.

File: proto/meson.build [hunk 1]
```diff
@@ -15,8 +15,10 @@ wayland_scanner_client = generator(
 )
 
 client_protocols = [
-    'wayfire-shell.xml',
-    'virtual-keyboard-unstable-v1.xml'
+	[wl_protocol_dir, 'stable/xdg-shell/xdg-shell.xml'],
```
Comment (ammen99) [inline proto/meson.build]: The indentation here seems wrong. We use 4 spaces.

File: src/main.cpp [hunk 2]
```diff
@@ -19,6 +21,7 @@ namespace wf
         int default_y = 100;
         int default_width = 800;
         int default_height = 400;
+		std::string anchor;
```
Comment (ammen99) [inline src/main.cpp]: Same as above, indentation. I noticed these are happening in many places, I won't list them all, please format at all places.

File: src/wayland-window.cpp [hunk 3]
```diff
@@ -44,6 +51,23 @@ namespace wf
         &registry_remove_object
     };
 
+    static void layer_shell_handle_configure(void *data, struct zwlr_layer_surface_v1 *zwlr_layer_surface,
```
Comment (ammen99) [inline src/wayland-window.cpp]: This line is too long. In general, we use a limit of 80 chars.

Sorry if some places in my code weren't properly formatted. I should also probably start providing an .editorconfig

Comment (ammen99) [reply src/wayland-window.cpp]: Also again I won't list this at all places, please check if you have over 80 chars in other places.

File: src/wayland-window.cpp [hunk 4]
```diff
@@ -74,18 +99,91 @@ namespace wf
         return instance;
     }
 
-    WaylandWindow::WaylandWindow(int x, int y, int width, int height)
-        : Gtk::Window()
+    uint32_t WaylandWindow::checkAnchorForWayfireShell(int width, int height, std::string anchor)
```
Comment (ammen99) [inline src/wayland-window.cpp]: Style: we use underscore_case not camelCase for methods (of course constructors have to match the type name)

File: src/wayland-window.cpp [hunk 5]
```diff
@@ -44,6 +51,23 @@ namespace wf
...
 
+    static void layer_shell_handle_configure(void *data, struct zwlr_layer_surface_v1 *zwlr_layer_surface,
+    		uint32_t serial, uint32_t width, uint32_t height)
+    {
+    	zwlr_layer_surface_v1_ack_configure(zwlr_layer_surface, serial);
+    }
+
+    static void layer_shell_handle_close(void *data, struct zwlr_layer_surface_v1 *surface)
+    {
+		zwlr_layer_surface_v1_destroy(surface);
```
Comment (ammen99) [inline src/wayland-window.cpp]: Shouldn't we quit the Gtk main loop in this case?

**Review by ammen99 — CHANGES_REQUESTED** [review]
Mostly style issues.

Do you think it would make sense to use https://github.com/wmww/gtk-layer-shell ? It would remove a lot of the needed boilerplate.

**Comment by masozzi** [comment]
@ammen99 Fixed all style issue.

> Do you think it would make sense to use https://github.com/wmww/gtk-layer-shell ? It would remove a lot of the needed boilerplate.

yes, but keep in mind that https://github.com/wmww/gtk-layer-shell is a library, so it would require one more dependency, but that's not a big deal.
Would you prefer a separate pull-request for this, or should i add it to this?

Also, there is still no way to close the keyboard on sway: do you think that adding a key to quit can be done? you have other ideas?

**Comment by ammen99** [comment]
I think we can just add gtk-layer-shell as a submodule if it isn't present on the system. Meson has support for this via the fallback parameter for dependencies. I would prefer to have this happen in this PR.

And yes, adding a key to close the keyboard can be done pretty easily by adding a new special keycode (if you look at that ugly hack which represents buttons)

However, IIRC we can add a GtkHeaderbar as a widget - in which case it will provide a close button. We need this only in the layer-shell case.

**Comment by AdrianVovk** [comment]
Adding a header bar for a close button is a bit overkill I think

**Comment by ammen99** [comment]
@AdrianVovk, what to do otherwise?

**Comment by AdrianVovk** [comment]
@ammen99 Just have a button and on click do window.destroy(). Adding in an entire headerbar just to close the osk is a bit much

**Comment by ammen99** [comment]
After some consideration I think the way forward is using only layer-shell with potentially optional behavior with wayfire-shell (wayfire shell will be redesigned to be used together with layer-shell). So I think the best would be a headerbar with close button, and this headerbar would also start an interactive drag in the future.

**Comment by masozzi** [comment]
@ammen99 
> After some consideration I think the way forward is using only layer-shell with potentially optional behavior with wayfire-shell (wayfire shell will be redesigned to be used together with layer-shell)

This means that layer-shell should have the priority over wf-shell? or the priority should remain to wf-shell, and in case it does not exist we go for layer-shell?

**Comment by ammen99** [comment]
layer-shell should have all priority. I am sorry for wasting your time in requesting that you support both. You can see what wayfire-shell will be like in the future: https://github.com/WayfireWM/wf-shell/blob/layer-shell/proto/wayfire-shell-unstable-v2.xml

So it will be all layer-shell and wayfire-shell will only provide support for the interactive drag (which I may implement later).

About gtk-layer-shell, you can add it as a subproject as I did in the layer-shell branch of wf-shell.

**Comment by masozzi** [comment]
@ammen99 i saw that you have include the commits from this PR into the `layer-shell-gtk` branch (plus some modifications), should i close this?

**Comment by ammen99** [comment]
@masozzi I was hoping you could test my changes, hopefully I didn't break anything. Then you can merge the PR and I'll merge this one (which will then include also my commits)

**Comment by masozzi** [comment]
@ammen99 i have tested your branch, happy to report that works.

**Review by ammen99 — APPROVED** [review]

**Comment by ammen99** [comment]
@masozzi Thanks for working on this initially and for testing :)

**Pull request merged**